### PR TITLE
Feature/multiframe

### PR DESCRIPTION
**Pull request by hackermd**
This PR adds a new class for reading individual frames from image datasets without having to load the entire pixel data element into memory.

### CODE REVIEW

File: tests/test_io.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            assert reader.number_of_frames == 1
+            frame = reader.read_frame(0)
+            assert isinstance(frame, np.ndarray)
+            assert frame.ndim == 2
+            assert frame.dtype == np.int16
+            assert frame.shape == (
+                reader.metadata.Rows,
+                reader.metadata.Columns,
+            )
+            np.testing.assert_array_equal(frame, pixel_array)
```
Comment (CPBridge) [inline tests/test_io.py:41]: There's a multi-frame ct image in the pydicom test data that would make a nice test also. It has just two frames.

```python
from pydicom.data import get_testdata_file

filename = get_testdata_file('eCT_Supplemental.dcm')
```

File: tests/test_utils.py [hunk 2]
```diff
@@ -100,44 +97,128 @@
...
+    # Patient
+    pytest.param(
+        dict(
+            coordinate=(5.0, 4.0),
+            image_position=(10.0, 60.0, 0.0),
+            image_orientation=(1.0, 0.0, 0.0, 0.0, 1.0, 0.0),
+            pixel_spacing=(0.5, 2.0),
+        ),
+        (12.5, 68.0, 0.0),
+    ),
```
Comment (CPBridge) [inline tests/test_utils.py:139]: Your current tests failed to find the error with the z coordinate calculation mentioned above because you always use 0.0 as the third element of `image_position`. I suggest adding some tests with non-zero values here. Also add some where the z component of the two cosine vectors is non-zero. E.g. a sagittal or coronal orientation. Here's an example for a sagittal image (confirmed with SimpleITK and my corrected version of the highdicom function above):

```python
pytest.param(
    dict(
        coordinate=(15.0, 87.0),
        image_position=(-45.0, 35.0, -90.0),
        image_orientation=(0.0, 1.0, 0.0, 0.0, 0.0, -1.0),  # axis-aligned sagittal orientation
        pixel_spacing=(0.5, 0.25),
    ),
    (-45.0, 42.5, -111.75),
)
```

File: src/highdicom/io.py [hunk 3]
```diff
@@ -0,0 +1,780 @@
...
+            'Number of frame items does not match specified Number of Frames.'
+        )
+    else:
+        basic_offset_table = offset_values
+
+    fp.seek(initial_position, 0)
+    return basic_offset_table
+
+
+def _build_icc_transform(metadata: Dataset) -> Union[ImageCmsTransform, None]:
```
Comment (CPBridge) [inline src/highdicom/io.py:229]: Consider contributing to pydicom.pixel_data_handlers.util instead of here?

File: src/highdicom/io.py [hunk 4]
```diff
@@ -0,0 +1,780 @@
...
+    # Skip Pixel Data element header (tag, VR, length)
+    pixel_data_element_value_offset = data_element_offset_to_value(
+        fp.is_implicit_VR, 'OB'
+    )
+    fp.seek(pixel_data_element_value_offset - 4, 1)
+    is_empty, offsets = get_frame_offsets(fp)
+    return offsets
+
+
+def _build_bot(fp: DicomFile, number_of_frames: int) -> List[int]:
```
Comment (CPBridge) [inline src/highdicom/io.py:141]: Consider contributing to pydicom.encaps instead of here?

**Review by CPBridge — CHANGES_REQUESTED** [review]
Some really nice stuff here, but I do have a two linked but distinct general concerns:

- Some of functionality added in this MR is pretty low-level and much closer to the scope of pydicom than highdicom. Things like the ICC colour mapping and even the framewise reading of images are quite low-level and generally applicable to a community beyond those likely to be working with highdicom. If the pydicom devs are interested, I think we should consider contributing some of these things upstream. 
- Some of functionality in here actually duplicates existing functionality in pydicom as far as I can tell, and may be far easier to maintain and expand if highdicom simply defers to pydicom on those matters (I have noted the precise locations in my other comments in more detail).

**Comment by hackermd** [comment]
> Some of functionality added in this MR is pretty low-level and much closer to the scope of pydicom than highdicom. Things like the ICC colour mapping and even the framewise reading of images are quite low-level and generally applicable to a community beyond those likely to be working with highdicom. If the pydicom devs are interested, I think we should consider contributing some of these things upstream

Fully agreed and I have already created an [issue](https://github.com/pydicom/pydicom/issues/1243). Unfortunately, this will require a significant amount of refactoring of the *pydicom* pixel data handlers. I would like to use our own implementation in the meanwhile, but update as soon as the relevant functionality becomes available in *pydicom*. The low level implementation details are abstracted in "private" functions and are not exposed by *highdicom*.

**Comment by hackermd** [comment]
> Some of functionality in here actually duplicates existing functionality in pydicom as far as I can tell, and may be far easier to maintain and expand if highdicom simply defers to pydicom on those matters

Yes, in some cases it was not possible to avoid copying code, since the relevant functionality is currently not exposed by *pydicom* via a reusable abstraction. We will see how this will ultimately get implemented in *pydicom*, but I would like to keep the high-level (opinionated) API definition in *highdicom* (`highdicom.io.ImageFileReader`). Moving forward, we will hopefully be able to replace more and more of the implementation using the *pydicom* pixel data handlers.

File: src/highdicom/content.py [hunk 5]
```diff
@@ -130,23 +130,28 @@ def __init__(
...
+            Offset of the first column and first row of the plane (frame) in
             pixels along the row and column direction of the total pixel matrix
             (only required if `coordinate_system` is ``"SLIDE"``)
 
         """
         super().__init__()
         item = Dataset()
+
+        def ds(num: float) -> float:
+            return float(str(num)[:16])
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/content.py:142]: Unfortunately, it doesn't. I just reported what I think is a bug to the *pydicom* developers via this [issue](https://github.com/pydicom/pydicom/issues/1264). I am afraid this is required for the time being.

File: src/highdicom/frame.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    JPEGBaseline,
+    JPEG2000Lossless,
+    RLELossless,
+)
+
+
+def encode_frame(
+    array: np.ndarray,
+    metadata: Dataset,
+) -> bytes:
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/frame.py:22]: Flake8 want's it this way (see [E1 error codes](https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes)). I found it weird at first too (and we are currently ignoring some of the error codes in `setup.cfg`), but I got used to it.

File: src/highdicom/frame.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    """Encodes pixel data of an individual frame.
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/frame.py:23]: Good point! What about adding the following check in `encode_frame()`?

```python
if (self.Rows * self.Columns * self.SamplesPerPixel) % 8 != 0:
    raise ValueError('Size of frame is not a multiple of 8 and can thus not be bit packed')
```

We would then also need to change the way the function is used by `highdicom.seg.sop.Segmentation`.

File: src/highdicom/frame.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    bytes
+        Pixel data (potentially compressed in case of encapsulated format
+        encoding, depending on the transfer snytax)
+
+    Raises
+    ------
+    ValueError
+        When transfer syntax is not supported.
+
+    """
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/frame.py:45]: See above :)

File: src/highdicom/frame.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            ),
+            JPEG2000Lossless: (
+                'jpeg2000',
+                {
+                    'tile_size': None,
+                    'num_resolutions': 1,
+                    'irreversible': False,
+                },
+            ),
+        }
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/frame.py:89]: Not a fan of global variables. What about adding a `encoding_params: Optional[Dict[str, Union[str, int, float, bool]]]` to allow users to pass transfer-syntax specific encoding parameters (as long as they are generic enough to potentially work with a different library other than *Pillow*).

File: src/highdicom/utils.py [hunk 10]
```diff
@@ -13,7 +13,6 @@ def tile_pixel_matrix(
     total_pixel_matrix_columns: int,
     rows: int,
     columns: int,
-    image_orientation: Tuple[float, float, float, float, float, float]
 ) -> Iterator[Tuple[int, int]]:
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py:16]: See comment above. Consistency among function definitions with the package would be a good idea. I will change it.

[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py]: Addressed with e7a21919d0e5c0a6768054c1e6aced52f2d9d92e

File: src/highdicom/frame.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/frame.py]: Addressed with e7a21919d0e5c0a6768054c1e6aced52f2d9d92e

File: src/highdicom/utils.py [hunk 11]
```diff
@@ -76,16 +65,15 @@ def compute_plane_position_tiled_full(
...
-        Relative one-based index value for a given frame along the row
+        One-based Row index value for a given frame (tile) along the column
         direction of the tiled Total pixel Matrix, which is defined by
-        the first triplet in `image_orientation`
+        the second triplet in `image_orientation` (values should be in the
+        range [1, *n*], where *n* is the number of tiles per column)
     column_index: int
-        Relative one-based index value for a given frame along the column
+        One-based Column index value for a given frame (tile) along the row
         direction of the the tiled Total Pixel Matrix, which is defined by
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py:72]: Addressed with a702343c857e49c6320eaeb14e8e12f544cd1afc

File: src/highdicom/utils.py [hunk 12]
```diff
@@ -76,16 +65,15 @@ def compute_plane_position_tiled_full(
     Parameters
     ----------
     row_index: int
-        Relative one-based index value for a given frame along the row
+        One-based Row index value for a given frame (tile) along the column
         direction of the tiled Total pixel Matrix, which is defined by
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py:67]: Addressed with 5d494f5a7ae4d3d6fa19d950e0d50f03bd9c83a2

File: src/highdicom/utils.py [hunk 13]
```diff
@@ -97,18 +85,20 @@ def compute_plane_position_tiled_full(
     columns: int
         Number of columns per Frame (tile)
     image_orientation: Tuple[float, float, float, float, float, float]
-        Cosines of row (first triplet: horizontal, left to right) and
-        column (second triplet: vertical, top to bottom) direction
-        for X, Y, and Z axis of the slide coordinate system
+        Cosines of the row direction (first triplet: horizontal, left to right,
+        increasing Column index) and the column direction (second triplet:
+        vertical, top to bottom, increasing Row index) direction for X, Y, and
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py:88]: Addressed with 3d1300c7a86dae6417ba253b9ae8890aa3f335b8

File: src/highdicom/utils.py [hunk 14]
```diff
@@ -122,23 +112,41 @@ def compute_plane_position_tiled_full(
...
+            '"slice_index", "slice_thickness", "spacing_between_slices"'
+        )
+    # These checks are needed for mypy to be able to determine the correct type
+    if (slice_index is not None and
+            slice_thickness is not None and
+            spacing_between_slices is not None):
+        z_offset = (
+            float(slice_index - 1) * slice_thickness +
+            float(slice_index - 1) * spacing_between_slices
+        )
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py:132]: Good catch! I wasn't aware of this detail. Addressed with d0fc87156a1477e47d1b70bc66ee4b3ed2b3c71e

File: src/highdicom/utils.py [hunk 15]
```diff
@@ -307,9 +319,7 @@ def map_pixel_into_coordinate_system(
 
     column_offset = float(coordinate[0])
     row_offset = float(coordinate[1])
-    pixel_matrix_coordinate = np.array([
-        [row_offset, column_offset, 0.0, 1.0]
-    ])
+    pixel_matrix_coordinate = np.array([[column_offset, row_offset, 0.0, 1.0]])
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py:320]: Thanks for the clarification. Coordinates in the slide coordinate system shall not be negative, but I forgot that this may occur in the patient coordinate system. I have removed the check.

Addressed with e67edb963fbf3ddb7bb261760768f279373ae068

File: src/highdicom/io.py [hunk 16]
```diff
@@ -0,0 +1,780 @@
...
+            self._fp.is_little_endian = transfer_syntax_uid.is_little_endian
+            self._fp.is_implicit_VR = transfer_syntax_uid.is_implicit_VR
+        except FileNotFoundError:
+            raise FileNotFoundError(f'File not found: "{self.filename}"')
+        except Exception:
+            raise OSError(
+                f'Could not open file for reading: "{self.filename}"'
+            )
+        try:
+            logger.info('read metadata elements')
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:388]: Addressed with 8cf94df8ead1486bab6aa6c7d38d860ce1da95fa

File: src/highdicom/io.py [hunk 17]
```diff
@@ -0,0 +1,780 @@
...
+    @property
+    def metadata(self) -> Dataset:
+        """pydicom.dataset.Dataset: Metadata"""
+        try:
+            return self._metadata
+        except AttributeError:
+            raise IOError('File has not been opened for reading.')
+
+    @property
+    def _pixels_per_frame(self) -> int:
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:452]: No because the function computes the length across all frames rather than a single frame.

File: src/highdicom/io.py [hunk 18]
```diff
@@ -0,0 +1,780 @@
...
+        """int: Number of pixels per frame"""
+        return np.prod([
+            self.metadata.Rows,
+            self.metadata.Columns,
+            self.metadata.SamplesPerPixel
+        ])
+
+    @property
+    def _bytes_per_frame_uncompressed(self) -> int:
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:461]: No because the function computes the length across all frames rather than a single frame.

File: src/highdicom/io.py [hunk 19]
```diff
@@ -0,0 +1,780 @@
...
+        logger.debug('build Basic Offset Table item')
+        basic_offset_table = _build_bot(
+            fp,
+            number_of_frames=number_of_frames
+        )
+
+    return basic_offset_table
+
+
+def _read_bot(fp: DicomFile) -> List[int]:
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:100]: It uses this function under the hood. It just makes sure the file pointer is positioned correctly.

File: src/highdicom/io.py [hunk 20]
```diff
@@ -0,0 +1,780 @@
...
+            #   are packed into 12.5 -> 13 bytes
+            return n_pixels // 8 + (n_pixels % 8 > 0)
+        else:
+            return n_pixels * bits_allocated // 8
+
+    def close(self):
+        """Closes file."""
+        self._fp.close()
+
+    def read_frame(self, index: int, correct_color: bool = True) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:477]: That's an elegant hack! I like it. That could serve as a lightweight workaround until we have figured out how to refactor the pydicom pixel data handlers.

Did you test this already?

File: tests/test_io.py (hunk 1 above)
Comment (hackermd) [reply tests/test_io.py:42]: Nice! Addressed with 4ba4eb5e0c80aa2351e989555729845f48c6361f

File: src/highdicom/io.py (hunk 3 above)
Comment (hackermd) [reply src/highdicom/io.py:229]: I already created an [issue](https://github.com/pydicom/pydicom/issues/1244) and would be happy to contribute that to *pydicom*. However, this function requires *Pillow*, which is optional in *pydicom*. @scaramallion any suggestions for a potential PR?

File: src/highdicom/io.py (hunk 4 above)
Comment (hackermd) [reply src/highdicom/io.py:141]: @scaramallion would you be open to including this in `pydicom.encaps`, for example as `pydicom.encaps.compute_frame_offsets()` (to match `pydicom.encaps.get_frame_offsets()`)?

File: src/highdicom/io.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/io.py:471]: It seems to work 👍  Addressed with 43c472877c2c1752ae5c71c9b6a5d829fd211e18

**Comment by hackermd** [comment]
@CPBridge I think I addressed most if not all your comments. Could you kindly take a second look?

File: src/highdicom/frame.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (CPBridge) [reply src/highdicom/frame.py:24]: Having spent quite a while carefully implementing this to make sure it does work with arrays that are not a multiple of 8 so that there are not limitations on the segmentations that can be encoded in the bit packed way (and therefore displayable in OHIF etc etc), I'm fairly reluctant to remove this capability without good reason. It would also break backwards compatibility

File: src/highdicom/frame.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (CPBridge) [reply src/highdicom/frame.py:90]: The intention of having it outside the scope of the function was just to keep to the code neater, as a "constant" (insofar as python has "constants"). Definitely not to encourage users to go tweaking the parameters.

Feel free to ignore and keep it as it is.

File: src/highdicom/io.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (CPBridge) [reply src/highdicom/io.py:478]: Sorry for delay, yes I tested it too but not very thoroughly.

File: src/highdicom/io.py (hunk 3 above)
Comment (CPBridge) [reply src/highdicom/io.py:220]: In that case, it would probably go into pydicom.pixel_data_handlers.pillow_handler

**Comment by CPBridge** [comment]
Nice, I think the only remaining thread is the one regarding the bit packing with frames that are not a multiple of 8

File: src/highdicom/frame.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/frame.py:46]: Addressed with 083f047224bb1077bd2e980cd3f7930bc12d7e07

File: src/highdicom/frame.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/frame.py:24]: Good point. I think that is a special enough case. I reverted the approach to encoding of segmentation frames for native transfer syntaxes (including bit packed), but use the `frame.encode_frame()` function for encapsulated transfer syntaxes. That should be a good compromise.

Addressed with eba8f2f41992237e699e835ecc473148277a463a

File: src/highdicom/frame.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (CPBridge) [reply src/highdicom/frame.py:46]: By this I meant that the input array is a 2D or 3D numpy array as described in the docstring, not checks on the size of the array along each of its dimensions

File: src/highdicom/frame.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (CPBridge) [reply src/highdicom/frame.py:24]: Looks good. However now there is a check in `_encode_frame` that the input array is a multiple of 8, which is unnecessarily restrictive. `_encode_frame` is perfectly capable of handling cases that are not a multiple of 8 - it just passes through to pydicom's `pack_bits`, which does the necessary padding and you don't need to worry about it.

The complicated situation is when you have *multiple* frames, each of which have  a number of pixels that does not divide by 8, and you want to bit pack them *together*. However since this function encodes just a single frame, this should not arise if used correctly.

File: src/highdicom/utils.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (hackermd) [reply src/highdicom/utils.py]: @CPBridge I refactored that quite a bit (see 6f543d476888a73fb8e710c97e8d1a2240f4a1ac) to facilitate reuse of the affine transformation matrix. In addition, I implemented the inverse transformation. It passes all the tests, but it would be great if you could test the inverse mapping (from frame of reference into pixel matrix) on a few more radiology images.

**Comment by CPBridge** [comment]
As requested, I have had a further look at the functions in the `highdicom.utils`.

I have created a new PR (#37) to fix some issues with the coordinate transformation functions.

I have a few other general comments about these functions:
1. I think it would make sense to put all the coordinate transform/spatial stuff in a dedicated `highdicom.spatial` or `highdicom.spatial_utils` module. 
2. Type hinting and input checks. Generally speaking i think we should aim for arguments to be type hinted as generically as possible. For example, where an image orientation is expected, I think the user should be able to pass a list of floats as well as a tuple of floats without their code failing type checking. Similarly a np.ndarray should be allowed. So the type hinting for many of these inputs could be `Union[Sequence[float], np.ndarray]`. Of course a downside is that this doesn't allow the type system to check the length of the input array, but I'd argue that a) this is basically impossible anyway since most image orientation/positions will come from pydicom datasets in the first place and will have to be coerced into Tuple[float, float, float] in a non type-safe way by user code anyway and b) since by no means all users will be using type hinting, unfortunately having type hinting cannot guarantee no runtime issues. Consequently I think we should have run-time checks on the length of the input arrays anyway where arrays of a specific length are expected (e.g. 3 for image positions, 6 for image orientations, 2 for pixel coordinates, 3 for physical coordinates). In this case the utility of having static checks on array length is limited.
3. Throughout the docs, it refers to pixel coordinates being one-based, but are you sure this is correct? The image position tag is described in the standard as being

> The x, y, and z coordinates of the upper left hand corner (center of the first voxel transmitted) of the image, in mm. 

Consequently, if pixel coordinates were one-based, we would expect an image with an image_position of [0.0, 0.0, 0.0] to a pixel coordinate of [1, 1], however it does not, instead it maps to [0, 0]:

```python
from highdicom import utils

utils.map_coordinate_into_pixel_matrix([0.0, 0.0, 0.0], [0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0, 1.0, 0.0], [1.0, 1.0])
# (0.0, 0.0, 0.0)
```

All the logic in the functions in the file is consistent with this definition of 0 based pixel indexing so I'm confused about why the documentation suggests otherwise. Am I missing something?

**Comment by hackermd** [comment]
> 1. I think it would make sense to put all the coordinate transform/spatial stuff in a dedicated `highdicom.spatial` or `highdicom.spatial_utils` module.

Agreed. I think that's a good idea.

> 2. Type hinting and input checks. Generally speaking i think we should aim for arguments to be type hinted as generically as possible. For example, where an image orientation is expected, I think the user should be able to pass a list of floats as well as a tuple of floats without their code failing type checking. Similarly a np.ndarray should be allowed. So the type hinting for many of these inputs could be Union[Sequence[float], np.ndarray]. Of course a downside is that this doesn't allow the type system to check the length of the input array, but I'd argue that a) this is basically impossible anyway since most image orientation/positions will come from pydicom datasets in the first place and will have to be coerced into Tuple[float, float, float] in a non type-safe way by user code anyway and b) since by no means all users will be using type hinting, unfortunately having type hinting cannot guarantee no runtime issues. Consequently I think we should have run-time checks on the length of the input arrays anyway where arrays of a specific length are expected (e.g. 3 for image positions, 6 for image orientations, 2 for pixel coordinates, 3 for physical coordinates). In this case the utility of having static checks on array length is limited.

Yes, I have stumbled upon this myself a few times and have been thinking about changing the type annotations of parameters. `Tuple[float, ...]` is definitely too restrictive and leads to issues with mypy, so I definitely agree to using `Sequence[float]` instead. However, I am not sure whether we would need support for `numpy.ndarray` for these utilities. The *pydicom* API returns lists for data element values. Using `numpy.ndarray` would also create some additional problems, because it would not specify the data type (`float` versus `int`), which is significant for some of the utilities. The arguments are only small 1D arrays with two or three elements and the caller could easily convert a NumPy array to a list using the `numpy.ndarray.tolist()` method. What do you think about using `Sequence[float]` as a compromise?

> 3. Throughout the docs, it refers to pixel coordinates being one-based, but are you sure this is correct?

The description of the *Row/Column Position In Total Image Pixel Matrix* attributes in [Table C.8.12.6.1-1. Plane Position (Slide) Macro Attributes](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.8.12.6.html#sect_C.8.12.6.1) states that "the row/column position of the top left pixel of the Total Pixel Matrix is 1". Therefore, I assumed that spatial coordinates in pixel unit defined relative to the Total Pixel Matrix start at 1.
However, the description of the *Graphic Data* attribute in [Table C.18.6-1. Spatial Coordinates Macro Attributes](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.6.html#table_C.18.6-1) states that "the origin at the TLHC of the TLHC pixel is 0.0\0.0, the BRHC of the TLHC pixel is 1.0\1.0, and the BRHC of the BRHC pixel is Columns\Rows. The values must be within the range 0\0 to Columns\Rows" (TLHC = Top Left Hand Corner, BRHC = Bottom Right Hand Corner). Therefore, I conclude that the floating-point values are zero-based. I think the confusing part here is that these values don't represent logical indices into a sequence/array, but rather represent spatial coordinates in a two-dimensional coordinate system confined by the the Total Pixel Matrix. Therefore, we should probably avoid the terms "one-based" and "zero-based" altogether and just state that values should be in the range `[0, Total Pixel Matrix Columns]` and `[0, Total Pixel Matrix Rows]`. The concept of the slice index is trickier, but we probably also want to think of it as a spatial coordinate rather than an index, even though that may not be technically correct. What do you think?

**Comment by hackermd** [comment]
I addressed comment `3.` with a9b10f0af8e1c86d1a13c2efcf652056d76a93ee

**Comment by hackermd** [comment]
Addressed comment `2.` with 02cf6d687faa7f198252cb6a4fb8b4bda1e775bb (not including `numpy.ndarray`)

**Comment by hackermd** [comment]
Addressed comment `1.` with ea5508caae21356ea2207cc4dc5675d897d3530a

**Comment by hackermd** [comment]
@CPBridge Regarding comment `2.`, for `image_orientation`, we should also consider allowing `Sequence[str]`, since the attribute as Value Representation DS and *pydicom* represents the values as `Sequence[str]`. I don't think mypy is able to figure that out without additional type hints, though. What do you think?

**Comment by hackermd** [comment]
@CPBridge As discussed recently, I also changed `apply_transform()` and `apply_inverse_transform()` to operate on an array of coordinates rather than on individual coordinates (see 19300da298f2eab852fb53f9daba36fb041f269f).

**Comment by CPBridge** [comment]
Sorry for the delay. Responding to my original comments above

1. Great, looks good
2. Yes I think it is a reasonable compromise to use Sequence[float]. That's a good point about Sequence[str], I think allowing Sequence[str] (and adding code to cast to float) would be a big win for usability.
3. I think I agree with your comment on one and zero based indexing. Seems like the standard is a little confusing on this point and we just have to do something sensible. Removing references to one-based coordinates (as you have done) I think will be clear enough for the library to behave as everyone will expect.

One very minor point I noticed when reading through the docstrings again:
- It is not very clear whether the `affine` parameter of `apply_inverse_transform` is supposed to be the forward transform or the inverse transform (i.e. whether the function inverts the transform before applying it). I think I would guess from the function name and docstring that it does invert the transform, which is not correct. Perhaps you could clarify the docstrings on this point, and perhaps say the `affine` input is an "inverse transform for example as returned by `build_inverse_transform`"?

Happy merging!

**Review by CPBridge — APPROVED** [review]

**Comment by hackermd** [comment]
Thanks @CPBridge! Regarding your last comment

> It is not very clear whether the affine parameter of apply_inverse_transform is supposed to be the forward transform or the inverse transform (i.e. whether the function inverts the transform before applying it). I think I would guess from the function name and docstring that it does invert the transform, which is not correct. Perhaps you could clarify the docstrings on this point, and perhaps say the affine input is an "inverse transform for example as returned by build_inverse_transform"?

This is a good point. Shall we rather remove `build_inverse_transform()` and create the inverse transform internally in `apply_inverse_transform()`? The advantage would be that the user would then only have to deal with the affine transformation matrix, which would be less error-prone. The disadvantage would be that the matrix inversion may be performed multiple times in a program, which would impact performance. However, I am not sure that performance penalty would matter. Which approach would you prefer?

**Comment by CPBridge** [comment]
I don't think it is necessary to remove the function, just be a bit clearer about what it does. Fundamentally the problem arises because we are using the word *inverse* in two different ways. One meaning is the reference space to image space transform, and the other is the inverse of any transform. I think I would favour maybe renaming the functions to something like this to sidestep the issue:

`build_transform` -> `build_image_to_reference_transform`
`apply_transform` -> `apply_image_to_reference_transform`
`build_inverse_transform` -> `build_reference_to_image_transform`
`apply_inverse_transform` -> `apply_reference_to_image_transform`

I think this makes it very clear how the functions should be used and most users would guess the correct way to use them just by looking at the function names. The downside is now the names are rather long, and shortening 'frame of reference' to 'reference' to avoid making it even worse is maybe a little unclear? @hackermd

**Comment by CPBridge** [comment]
Yeah that's probably even better

**Comment by hackermd** [comment]
@CPBridge addressed via 5e4c3c7535763ddec2c7ff36b5e8edfd46aa3c8c

**Comment by CPBridge** [comment]
Looks good
